source/logrepl/cdc.go

```diff
@@ -18,83 +18,117 @@ import (
+func NewCDCIterator(ctx context.Context, pgconf *pgconn.Config, c CDCConfig) (*CDCIterator, error) {
+	conn, err := pgconn.ConnectConfig(ctx, withReplication(pgconf))
+	if err != nil {
+		return nil, fmt.Errorf("could not establish replication connection: %w", err)
 	}
 
-	err := i.attachSubscription(connPool.Config().ConnConfig.Config)
-	if err != nil {
-		return nil, fmt.Errorf("failed to setup subscription: %w", err)
+	if err := internal.CreatePublication(
+		ctx,
+		conn,
+		c.PublicationName,
+		internal.CreatePublicationOptions{Tables: c.Tables},
+	); err != nil {
+		// If creating the publication fails with code 42710, this means
+		// the publication already exists.
+		if !internal.IsPgDuplicateErr(err) {
+			return nil, err
+		}
+
+		sdk.Logger(ctx).Warn().
+			Msgf("Publication %q already exists.", c.PublicationName)
 	}
 
-	go i.listen(ctx)
+	records := make(chan sdk.Record)
+
+	sub, err := internal.CreateSubscription(
+		ctx,
+		conn,
+		c.SlotName,
+		c.PublicationName,
+		c.Tables,
+		c.LSN,
+		NewCDCHandler(internal.NewRelationSet(), c.TableKeys, records).Handle,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize subscription: %w", err)
+	}
 
-	return i, nil
+	return &CDCIterator{
+		config:  c,
+		records: records,
+		pgconn:  conn,
+		sub:     sub,
+	}, nil
 }
 
-// listen should be called in a goroutine. It starts the subscription and keeps
-// it running until the subscription is stopped or the context is canceled.
-func (i *CDCIterator) listen(ctx context.Context) {
+// StartSubscriber starts the logical replication service in the background.
+// Blocks until the subscription becomes ready.
+func (i *CDCIterator) StartSubscriber(ctx context.Context) error {
 	sdk.Logger(ctx).Info().
 		Str("slot", i.config.SlotName).
 		Str("publication", i.config.PublicationName).
-		Msg("starting logical replication")
+		Msg("Starting logical replication")
 
-	err := i.sub.Start(ctx)
-	if err != nil {
-		// log it to be safe we don't miss the error, but use warn level
-		// because the error will most probably be still propagated to Conduit
-		// and might be recovered from
-		sdk.Logger(ctx).Warn().Err(err).Msg("subscription returned an error")
-	}
+	go func() {
+		if err := i.sub.Start(ctx); err != nil {
+			sdk.Logger(ctx).Error().
+				Err(err).
+				Msg("replication exited with an error")
+		}
+	}()
+
+	<-i.sub.Ready()
+	i.subStarted.Store(true)
+
+	sdk.Logger(ctx).Info().
+		Str("slot", i.config.SlotName).
+		Str("publication", i.config.PublicationName).
+		Msg("Logical replication started")
+
+	return nil
 }
 
 // Next returns the next record retrieved from the subscription. This call will
 // block until either a record is returned from the subscription, the
 // subscription stops because of an error or the context gets canceled.
+// Returns error when the subscription has been started.
 func (i *CDCIterator) Next(ctx context.Context) (sdk.Record, error) {
+	if started := i.subStarted.Load(); !started {
```

Comment:
Maybe `!i.subStarted.Load()`?

Reply:
sorry I changed this code, I can simplify it.

Reply:
fixed here f8b0599